core/new-gui/src/app/workspace/service/drag-drop/drag-drop.service.ts

```diff
@@ -334,105 +342,174 @@ export class DragDropService {
+        if (array[mid] != null) {
+          low = mid + 1;
+        } else { high = mid; }
+      }
+      array.splice(low, array.length - low);
+    };
+
+    // distances to closest operators sorted least to greatest
+    const inputMinDists: number[] = [];
+    const outputMinDists: number[] = [];
+    // closest operators sorted least to greatest by distance
+    const inputOps: (OperatorPredicate | null)[] = [];
+    const outputOps: (OperatorPredicate | null)[] = [];
+
+    // fill input/output arrays with default values
+    for (let i = 0; i < numInputOps; i++) {
+      inputMinDists.push(Number.MAX_VALUE);
+      inputOps.push(null);
+    }
+    for (let i = 0; i < numOutputOps; i++) {
+      outputMinDists.push(Number.MAX_VALUE);
+      outputOps.push(null);
+    }
+
+    // for each operator, check if in range/has free ports/is on the right side/is closer than prev closest ops/
     operatorList.forEach(operator => {
       const operatorPosition = this.workflowActionService.getJointGraphWrapper().getOperatorPosition(operator.operatorID);
       const distanceFromCurrentOperator = Math.sqrt((mouseCoordinate.x - operatorPosition.x) ** 2
-      + (mouseCoordinate.y - operatorPosition.y) ** 2);
-
-      if (distanceFromCurrentOperator < minDistance && distanceFromCurrentOperator < DragDropService.SUGGESTION_DISTANCE_THRESHOLD) {
-
-        const existPortCountFromSource = operatorLinks
-          .filter(link => link.source.operatorID === operator.operatorID).length;
-        const existPortCountFromTarget = operatorLinks
-          .filter(link => link.target.operatorID === operator.operatorID).length;
-
-        // check left or right, if current operator can add input or output ports, and the operator
-        //  we are checking to suggest has output/input ports available to add
-        if ((operatorPosition.x < mouseCoordinate.x && curruntOperator.inputPorts.length > 0 &&
-          existPortCountFromSource !== operator.outputPorts.length) ||
-            (operatorPosition.x >= mouseCoordinate.x && curruntOperator.outputPorts.length > 0 &&
-              existPortCountFromTarget !== operator.inputPorts.length)) {
-
-            // check if the dragging operator is on the left or right of the selected operator
-            this.isSuggestionOnLeft = operatorPosition.x < mouseCoordinate.x ? true : false;
+        + (mouseCoordinate.y - operatorPosition.y) ** 2);
+
+      if (distanceFromCurrentOperator < DragDropService.SUGGESTION_DISTANCE_THRESHOLD) {
+
+        // if it can be an input op
+        if (operatorPosition.x < mouseCoordinate.x
+          && distanceFromCurrentOperator < inputMinDists[numInputOps - 1] && hasFreeOutputPorts(operator)) {
+          const operatorIndex = sortedIndex(inputMinDists, distanceFromCurrentOperator);
+          // splice in new min-dist and new op
+          inputMinDists.splice(operatorIndex, 0, distanceFromCurrentOperator);
```

Comment:
These lines form the main recommendation mechanism. This mechanism works very well and feels natural to the users. 
However, I feel how this mechanism works is not well documented. The logic might look obvious to you, but it took me quite some time to understand it. I believe it is important to let other people easily understand a piece of code, as the people who maintain the code in the future might not be the original author.  

From what I understand, you are maintaining a sorted array of top n operators (n being number of input/outputs) that are the closest to the current operator. It would be nice if you are add such kind of high level description.
Moreover, I feel it is most natural to use a priority queue for this purpose. Rather than hand-writing the primitive functions like inserting, deleting, and finding sorted index, I prefer to use a library, which makes the code shorter and eaiser to understand. You can try https://github.com/mourner/tinyqueue
The library allows you to store object and use custom comparator. So I would create a priority queue of tuple `[OperatorPredicate, number]` and pass a comparator. In this way, this function would be shorter and easier to read.

Reply:
On review, the improvements probably weren't worth it. Too small compared to the "performance concern" functions. I will be refactoring this.